### PR TITLE
Fixed issue #226, where anonymous classes were not considered a valid resolution context

### PR DESCRIPTION
**Pull request by staslev**
@ftomassetti , let me know how it looks.

See also #226.

### CODE REVIEW

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+import com.google.common.base.Preconditions;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A parser context for an object creation node.
+ */
+// based on JavaParserTypeDeclarationAdapter
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: To be removed

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import com.github.javaparser.symbolsolver.reflectionmodel.ReflectionClassDeclaration;
+import com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic;
+import com.google.common.base.Preconditions;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A parser context for an object creation node.
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: It is more a symbol resolution context than a parser context

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+// based on JavaParserTypeDeclarationAdapter
+public class AnonymousClassDeclarationContext extends AbstractJavaParserContext<ObjectCreationExpr> {
+
+  private final JavaParserAnonymousClassDeclaration myDeclaration =
+      new JavaParserAnonymousClassDeclaration(wrappedNode, typeSolver);
+
+  public AnonymousClassDeclarationContext(ObjectCreationExpr node, TypeSolver typeSolver) {
+    super(node, typeSolver);
+    Preconditions.checkArgument(node.getAnonymousClassBody().isPresent(),
+                                "An anonymous class must have a body present");
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: "An anonymous class must body must be present"
or
"An anonymous class must have a body"

Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: +1 for using Preconditions

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  @Override
+  public SymbolReference<MethodDeclaration> solveMethod(String name,
+                                                        List<Type> argumentsTypes,
+                                                        boolean staticOnly,
+                                                        TypeSolver typeSolver) {
+    List<MethodDeclaration> candidateMethods =
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: At some point we may want to refactor the common parts of the different implementation of `solveMethod` in one utility class

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    return MethodResolutionLogic.findMostApplicable(candidateMethods,
+                                                    name,
+                                                    argumentsTypes,
+                                                    typeSolver);
+  }
+
+  @Override
+  public SymbolReference<TypeDeclaration> solveType(String name, TypeSolver typeSolver) {
+
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: Please remove this empty line

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            .stream()
+            .filter(internalType -> name.startsWith(String.format("%s.", internalType.getName())))
+            .findFirst()
+            .map(internalType ->
+                     JavaParserFactory
+                         .getContext(internalType, typeSolver)
+                         .solveType(name.substring(internalType.getName().getId().length() + 1),
+                                    typeSolver));
+
+    // fancy way of saying "use exactMatch if present, otherwise, use recursiveMatch"
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: Then we can avoid calculating the recursiveMatch, if not needed. Not sure how "lazy" the evaluation of that stream is.

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnonymousClassDeclaration.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+/**
+ * An anonymous class declaration representation.
+ */
+public class JavaParserAnonymousClassDeclaration extends AbstractClassDeclaration {
+
+  private final TypeSolver typeSolver;
+  private final ObjectCreationExpr wrappedNode;
+  private final TypeDeclaration superTypeDeclaration;
+  private final String name = "Anonymous-" + UUID.randomUUID();
```
Comment (ftomassetti) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnonymousClassDeclaration.java]: Mmm, I am not sure I am happy with this. But I cannot think of a better alternative.

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 2 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: 👍

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 1 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: 👍

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 3 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: 👍

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 5 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: 👍

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 4 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: Yes, at the moment it seems that the `JavaParserTypeDeclarationAdapter` class has a lot in common with the `AnonymousClassDeclarationContext`, but the latter can't use the former as is due to incompatible types. 

Definitely something worth looking at at some point.

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java (hunk 6 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/contexts/AnonymousClassDeclarationContext.java]: I suspect that in order to make it truly lazy I'd have to get the whole block into an `orElseGet(...)`, which lacks elegancy, in which case I can just resort to an `if` statement.

File: java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnonymousClassDeclaration.java (hunk 7 above)
Comment (staslev) [inline java-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/declarations/JavaParserAnonymousClassDeclaration.java]: No happiness here.

**Comment by staslev** [comment]
Revised according to comments.

Let me know if I missed anything.

**Comment by ftomassetti** [comment]
Thanks, it looks good

**Pull request merged**